src/polysolve/nonlinear/Solver.hpp

```diff
@@ -59,30 +62,35 @@ namespace polysolve::nonlinear
                const double characteristic_length,
                spdlog::logger &logger);
 
+        virtual ~Solver() = default;
+
+        const TCriteria &get_stop_criteria() { return m_stop; }
+        void set_stop_criteria(const TCriteria &s) { m_stop = s; }
+
+        const TCriteria &criteria() { return m_current; }
+        const Status &status() { return m_status; }
+
         void set_strategies_iterations(const json &solver_params);
 
-        virtual double compute_grad_norm(const Eigen::VectorXd &x, const Eigen::VectorXd &grad) const;
+        virtual double compute_grad_norm(const TVector &x, const TVector &grad) const;
 
         void set_line_search(const json &params);
 
-        void minimize(Problem &objFunc, TVector &x) override;
+        void minimize(Problem &objFunc, TVector &x);
 
         const json &get_info() const { return solver_info; }
 
         ErrorCode error_code() const { return m_error_code; }
 
-        const typename Superclass::TCriteria &getStopCriteria() { return this->m_stop; }
-        // setStopCriteria already in ISolver
-
         bool converged() const
         {
-            return this->m_status == cppoptlib::Status::XDeltaTolerance
-                   || this->m_status == cppoptlib::Status::FDeltaTolerance
-                   || this->m_status == cppoptlib::Status::GradNormTolerance;
+            return m_status == Status::XDeltaTolerance
```

Comment:
maybe add a static function in the Criteria file, so if we extend it it is in the same place